### ISSUE TITLE
Customizing iterable

### ISSUE THREAD

**Issue by tgriesser** [transduce/transduce#39]
So I'm wondering about the `iterable` implementation... I'd sort of like to be able to extend it so that as opposed to defaulting to `ObjectIterable` I could tweak it to only use `ObjectIterable` if it passes lodash's `isPlainObject` check (otherwise it's likely an error)...

Something like

```js
import {iterable as _iterable} from 'transduce'

export function iterable(value) {
   return _iterable(value, (ObjectIterable) => {
     if (isPlainObject(value)) {
       return new ObjectIterable(value)
     }
     throw new Error(`Unable to iterate: ${value}`
   })
}
```

Thoughts?

**Comment by kevinbeaty** [comment]
Are you suggesting a way to extend `iterable` and replace the implementation used by `transduce` internally?  Or adding an `isPlainObject` guard within the `iterable` implementation?

I've actually been wondering lately if I should add a `lodash-node` dependency to the project and import utilities as needed.  I've played with this in [`underarm`](https://github.com/kevinbeaty/underarm).   If we did this, we could add the `isPlainObject` guard and also replace all the `is*` checks in `util` with imports from `lodash-node`.  People using browserify or webpack for builds could also use lodash-node and only bundle the util functions once (or probably exclude the `lodash` functions from the bundle and rely on a global installation).

Also, treating strings as arrays in the iterable implementation may be problematic.  I know lodash has many workarounds for string iteration, but not sure if we could make use of anything there.

**Comment by tgriesser** [comment]
I was suggesting a way to extend/replace rather than add the `isPlainObject` guard which IIRC is a little hefty.

I kind of like the fact that there are few to no dependencies, I was actually going to ask about splitting out transduce async separately but hadn't thought about it enough. Most of the `is*` checks can also be accomplished with `typeof` checks (which is probably faster than an extra function call but I guess that's a micro optimization).

Also strings implement `Symbol.iterator` in ES6 anyway so this seems like the right behavior.

**Comment by kevinbeaty** [comment]
OK.

> I was suggesting a way to extend/replace rather than add the isPlainObject guard which IIRC is a little hefty.

We could do something like [redispatch](https://github.com/kevinbeaty/redispatch), which is what I did to extend `iterator` in `underarm`. Open to other options as well.

> I kind of like the fact that there are few to no dependencies,

I do too, usually.  I go back and forth on this. Sometimes I think the "no-dependencies" argument is a little silly, but I do often feel the same way.

> I was actually going to ask about splitting out transduce async separately but hadn't thought about it enough

Could, but I'm not really sure where to draw the line.   I used to have separate `transduce-array`, `transduce-async`, etc. but merged them later.

Any particular reason you don't want the async stuff?  Is it just the added dependency?  Now that we are using ES6, I could just assume a global Promise (using babeljs pollyfill for Node.js) and remove the dependency on `any-promise`.

**Comment by kevinbeaty** [comment]
We could separate out all packages into `transduce-contrib` and leave just `core` and `transducers` here.  Bring things across as they are proven.  See #40 for discussion.

As far as extending `iterable`, I'm currently hesitant on adding that complexity.  I know it is probably an error if it's not a plain object, but the user did ask for an `iterable` and got one as the last option.  If there were an easy way to check "plain object" vs any other object, I would consider adding it to the implementation.  But I think it's more trouble than it's worth.

Feel free to convince me otherwise.

**Comment by tgriesser** [comment]
> Could, but I'm not really sure where to draw the line. I used to have separate transduce-array, transduce-async, etc. but merged them later.

Yeah let's keep them consolidated for now, I'll think on it a little more - it was mainly just the added dependency but I see how any-promise works and it's pretty straightforward and small, let's keep it as it is now.

> As far as extending iterable, I'm currently hesitant on adding that complexity. I know it is probably an error if it's not a plain object, but the user did ask for an iterable and got one as the last option. If there were an easy way to check "plain object" vs any other object, I would consider adding it to the implementation. But I think it's more trouble than it's worth.

Yeah I guess this would be something that could be re-implemented quite easily myself - what do you think about exposing the Array/Function/Object Iterator/Transformer classes instead?

Basically what I'm thinking of doing is wrapping `transduce` in another lib and adding features/tweaks that I'm sort of experimenting with that may or may not make sense to eventually upstream once refined - for example I've got an iterator which recursively flattens out iterators/iterables which I've found to be pretty useful but might not make sense here.

**Comment by kevinbeaty** [comment]
> Yeah let's keep them consolidated for now,

OK. Sounds good.

> what do you think about exposing the Array/Function/Object Iterator/Transformer classes instead

I'd be OK with this.  Let's leave it out of the readme for now as a feature to be discovered, just in case we change our mind about supporting this in the future. 

> Basically what I'm thinking of doing is wrapping transduce in another lib

Cool.  I'm doing something similar with my own library (although this one was extracted from it so I went the other way). 

One thing I've been mulling over is the "identity" of this library. One thought is to let it be a repository of experimental features (mentioned in #28 ), another is a minimal ES6 implementation of proven features but as a basis for experimentation in external libraries (ala #40 ).   Basically, what is the advantage of this library vs contributing to the others (transducers-js, transducers.js, now Ramda)?  To be honest, I considered deprecating this library when the Ramda PR was accepted, but then others (mostly you) have indicated interest.

Any thoughts?

**Comment by tgriesser** [comment]
> Let's leave it out of the readme for now as a feature to be discovered, just in case we change our mind about supporting this in the future.

Perfect

>  Basically, what is the advantage of this library vs contributing to the others (transducers-js, transducers.js, now Ramda)?

Well `transducers-js` doesn't take PR's and is intentionally lacking in features (it's meant as a baseline reference implementation of transducers) - James has been pretty busy lately with other things and so `transducers.js` sort of fell behind a bit (which is how I originally came across this lib as I was digging for other implementations). I know there's also another implementation here: https://github.com/Gozala/transducers but yeah I think it's great to have different libraries experimenting with this, finding different patterns and such as it's still a fairly new idea and ES6 iterators haven't quite gained a ton of traction yet.

I also don't need the power of Ramda for the cases I'm working on, I think mainly what I'm after is a very lightweight (lightweight as in function calls, not size) library which makes it simpler to compose iterables, which none of the major utility libraries are capable of.

>  To be honest, I considered deprecating this library when the Ramda PR was accepted, but then others (mostly you) have indicated interest.

If you're considering deprecating (if you don't think it's worth it for you personally to keep working on it) let me know, I'd gladly take over or help out!

**Comment by kevinbeaty** [comment]
> I'd gladly take over or help out!

I'm not quite ready to give it up yet, but would be very interested in collaboration.  Just sent you an invite. 

> I think mainly what I'm after is a very lightweight (lightweight as in function calls, not size) library which makes it simpler to compose iterables, which none of the major utility libraries are capable of.

Sounds good to me.

Feel free to create any experimental branches in `transduce` and PR to master at will.   I think we should PR everything to master and approve each others for now.  That seems to work well in other projects.

**Comment by tgriesser** [comment]
> I think we should PR everything to master and approve each others for now.

Yep, sounds great! I think that's the ideal for open source - it's great to have someone else as a sounding board for different things
